modules/dag-l0/src/main/scala/org/tessellation/dag/l0/domain/snapshot/programs/SnapshotBinaryFeeCalculator.scala

```diff
@@ -0,0 +1,51 @@
+package org.tessellation.dag.l0.domain.snapshot.programs
+
+import cats.MonadThrow
+import cats.syntax.applicative._
+import cats.syntax.either._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+
+import scala.collection.immutable.SortedMap
+
+import org.tessellation.dag.l0.infrastructure.snapshot.StateChannelEvent
+import org.tessellation.node.shared.domain.statechannel.FeeCalculatorConfig
+import org.tessellation.schema.currencyMessage.fetchStakingBalance
+import org.tessellation.schema.{GlobalSnapshotInfo, SnapshotOrdinal}
+
+import eu.timepit.refined.types.numeric.NonNegLong
+
+trait SnapshotBinaryFeeCalculator {
+  def calculateFee[F[_]: MonadThrow](event: StateChannelEvent, info: GlobalSnapshotInfo, ordinal: SnapshotOrdinal): F[NonNegLong]
+}
+
+object SnapshotBinaryFeeCalculator {
+  def make(configs: SortedMap[SnapshotOrdinal, FeeCalculatorConfig]): SnapshotBinaryFeeCalculator =
+    new SnapshotBinaryFeeCalculator {
+      override def calculateFee[F[_]: MonadThrow](
+        event: StateChannelEvent,
+        info: GlobalSnapshotInfo,
+        ordinal: SnapshotOrdinal
+      ): F[NonNegLong] =
+        event.address
+          .pure[F]
+          .map(fetchStakingBalance(_, info))
+          .flatMap { balance =>
+            val config = FeeCalculatorConfig.getByOrdinal(configs)(ordinal)
+            val binary = event.snapshotBinary.value
+            val kbytes = BigDecimal(binary.content.length) / 1024
+            val workAmount = kbytes * config.computationalCost.value
+            val workMultiplier = 1.0 / (1.0 + BigDecimal(balance.value.value) * config.stakingWeight.value)
+
+            val feePerKb = BigDecimal(binary.fee.value.value) * kbytes
+            val optionalTip = BigDecimal(0).max(feePerKb - config.baseFee.value)
+
+            val fee = BigDecimal(config.baseFee.value) * workAmount * workMultiplier + optionalTip
```

Comment:
I've already seen this in some @TheMMaciek PR. Isn't it a duplication?

Reply:
just to be clear, the main difference between that and this is the optionalTip portion + the sorting vs. valid/invalid, any parts that are in common can be use same code if possible

Reply:
@ryle-goehausen if so, it looks like we can reuse this one with optional tip set to None and take top1 for validity. is it correct? Result of this validation can be simplified to valid/invalid so we would have the same code for validation but use the result in a slightly different way (rich output vs valid/invalid)

Reply:
@marcinwadon @ryle-goehausen @TheMMaciek 
I can refactor my code to use the `FeeCalculator`. I see `tip` parameter is not used so I suggest replacing it with `feePerKb` so that it can calculate the tip itself.